Spec change for auto-props (#4549)
dotnet/csharplang: spec/classes.md

### Automatically implemented properties

An automatically implemented property (or ***auto-property*** for short), is a non-abstract non-extern property with at least:

1. one semicolon-only body

> 2. usage of the `field` contextual keyword ([Keywords](lexical-structure.md#keywords)) within the `property_body` of the `property`. The `field` identifier is only considered the `field` contextual when there is no symbol named `field` in scope at that location. For the purposes of simplicity, from this point on any usages of `field` in the specification description will assume there is no `field` symbol in scope at that location.

Reviewer: ```suggestion
2. usage of the `field` contextual keyword ([Keywords](lexical-structure.md#keywords)) within the `property_body` of the `property`. The `field` identifier is only considered the `field` contextual keyword when there is no symbol named `field` in scope at that location. For the purposes of simplicity, from this point on any usages of `field` in the specification description will assume there is no `field` symbol in scope at that location.
```

Auto-properties must have a get accessor and can optionally have a set accessor. Because an expression-bodied property defines a get accessor, if the expression-bodied property refers to `field`, it satisfies the requirements and is considered an auto-property.

When a property is specified as an auto-property, a hidden, unnamed, backing field is automatically available for the property, and any semicolon-only accessors are implemented to read from and write to that backing field. The hidden backing field is in scope in all accessors and can be referenced directly using `field`.  Because the field is unnamed, it cannot be used in a `nameof` expression.

>  If the auto-property has no set accessor, and does not use `field`, the backing field is considered `readonly` ([Readonly fields](classes.md#readonly-fields)). Just like a `readonly` field, a getter-only auto-property can also be assigned to in the body of a constructor of the enclosing class. Such an assignment assigns directly to the readonly backing field of the property.

Reviewer: That doesn't seem obvious. Can't I have an auto-incrementing get-only property?

Author: this was tweaked to separate out the concepts of 'readonly' (which only occurs for a `get;`-only autoprop) and 'constructor can assign directly to field'.  Can you recheck?

### Automatically implemented properties

An automatically implemented property (or ***auto-property*** for short), is a non-abstract non-extern property with either or both:

1. an accessor with semicolon-only body
2. usage of the `field` contextual keyword ([Keywords](lexical-structure.md#keywords)) within the accessors or expression body of the property. The `field` identifier is only considered the `field` keyword when there is no existing symbol named `field` in scope at that location.

Auto-properties must have a get accessor and can optionally have a set accessor. Because an expression-bodied property defines a get accessor, if the expression-bodied property refers to the `field` keyword, it satisfies the requirements and is considered an auto-property.

> When a property is specified as an auto-property, a hidden, unnamed, backing field is automatically available for the property, and any semicolon-only accessors are implemented to read from and write to that backing field. The backing field can be referenced directly using the `field` keyword within all accessors and within the property expression body. Because the field is unnamed, it cannot be used in a `nameof` expression.

Reviewer 2: ```suggestion
When a property is specified as an auto-property, a hidden, unnamed backing field is automatically available for the property, and any semicolon-only accessors are implemented to read from (if a `get` accessor) or write to (if a `set` accessor) that backing field. The backing field can be referenced directly using the `field` keyword within all accessors and within the property expression body. Because the field is unnamed, it cannot be used in a `nameof` expression.
```

Reviewer 2: (I know those parens are not in the original, but what the original says is kinda ambiguous at the moment).

Author: Yeah, i'm ok with keeping as is given the original.  We can see what LDM thinks on this.   but i think it's fairly clear contextually :)

> When a property is specified as an auto-property, a hidden, unnamed, backing field is automatically available for the property, and any semicolon-only accessors are implemented to read from and write to that backing field. The backing field is in scope in all accessors, the `property_initializer`, the property expression body, and can be referenced directly using the `field` keyword. Because the field is unnamed, it cannot be used in a `nameof` expression.

Reviewer: Should probably state what is the type of the backing field ;-)

Author: interesting.  we never seemed to need that before :)

> If the auto-property does not have a set accessor, the backing field can still be assigned to in the body of a constructor of the enclosing class. Such an assignment assigns directly to the readonly backing field of the property.

Reviewer 2: ```suggestion
Just like a `readonly` field, a getter-only auto-property can also be assigned to in the body of a constructor of the enclosing class. Such an assignment assigns directly to the readonly backing field of the property.
```

I much prefer the original wording here.

Author: the problem here is what does 'getter only' mean.  It doesn't mean it has an explicit 'get accessor', as `=>` auto-props would count here.  So really, all we're saying is that even if there is no set-accessor, the auto-prop is still assignable in the constructor.

}
```
is equivalent to the following declaration:
```csharp
public class ReadOnlyPoint
{
    private readonly int __x;
    private readonly int __y;
    public int X { get { return __x; } }
    public int Y { get { return __y; } }
    public ReadOnlyPoint(int x, int y) { __x = x; __y = y; }
}
```

Notice that the assignments to the readonly field are legal, because they occur within the constructor.

The following example:
```csharp
// No 'field' symbol in scope.
public class Point
{
    public int X { get; set; }
    public int Y { get; set; }
}
```
is equivalent to the following declaration:
```csharp
// No 'field' symbol in scope.
public class Point
{
    public int X { get { return field; } set { field = value; } }
    public int Y { get { return field; } set { field = value; } }
}
```
which is equivalent to:
```csharp
// No 'field' symbol in scope.
public class Point
{
    private int __x;
    private int __y;
    public int X { get { return __x; } set { __x = value; } }
    public int Y { get { return __y; } set { __y = value; } }
}
```

The following example:
```csharp
// No 'field' symbol in scope.
public class LazyInit
{

>     public string Value => field ??= ComputeValue();

Reviewer 3: Happy to see we don't need `auto` modifier :)

Is there any special meaning for `@field`?

Author: yes.  it would no longer be a keyword (contextual or otherwise) so would not count.  this falls out from our definitions around keywords/identifiers and whatnot.

Author: > Happy to see we don't need auto modifier :)

LDM still to decide on this.  it will come back to if we think it is critical to be able to tell if somethign is an autoprop without having to do examination of the bodies.  My perspective is that it is not critical and it's acceptable to have no top-level sigil indicating this information.

Notice that the assignments to the readonly field are legal, because they occur within the constructor.

The following example:
```csharp
// No 'field' symbol in scope.
public class Point
{

> 	public int X { get; set; }

Reviewer: nit: indent in code snippets look unusual

	public int Y { get; set; }
}
```
is equivalent to the following declaration:
```csharp
// No 'field' symbol in scope.
public class Point
{
	public int X { get { return field; } set { field = value; } }
	public int Y { get { return field; } set { field = value; } }
}
```
which is equivalent to:
```csharp
// No 'field' symbol in scope.
public class Point
{
    private int __x;
    private int __y;
    public int X { get { return __x; } set { __x = value; } }
    public int Y { get { return __y; } set { __y = value; } }
}
```

The following example:
```csharp
// No 'field' symbol in scope.
public class LazyInit
{

> 	public string Value => field ??= ComputeValue();

Reviewer: nit: this indent is an eye sore. Also below (line 2975)

Author: yup.  all whitespace was fixed.
